File: packages/sdk-go/pkg/parallax/agent_service.go
package parallax

import (
	"context"
	"fmt"

	"parallax/sdk-go/generated"
	"go.uber.org/zap"
)

// agentService implements AgentService
type agentService struct {
	client *Client
	logger *zap.Logger
	leases map[string]string
}

// Register registers a new agent
func (s *agentService) Register(ctx context.Context, agent *AgentInfo) error {
	s.logger.Info("Registering agent",
		zap.String("id", agent.ID),
		zap.String("name", agent.Name),
		zap.Strings("capabilities", agent.Capabilities),
	)
	
	if agent.ID == "" {
		agent.ID = generateID()
	}
	
	if agent.Status == "" {
		agent.Status = AgentStatusActive
	}
	
	if agent.Confidence == 0 {
		agent.Confidence = 0.8
	}
	
	agent.LastSeen = timeNow()

	labels := map[string]string{}
	for key, value := range agent.Metadata {
		labels[key] = value
	}

	registration := &generated.AgentRegistration{
		Id:           agent.ID,
		Name:         agent.Name,
		Endpoint:     agent.Endpoint,
		Capabilities: agent.Capabilities,
		Metadata: &generated.AgentRegistration_Metadata{
			Version:          agent.Metadata["version"],
			Region:           agent.Metadata["region"],
			Labels:           labels,
			DefaultConfidence: agent.Confidence,
		},
	}

	client := generated.NewRegistryClient(s.client.conn)
	response, err := client.Register(ctx, &generated.RegisterRequest{
		Agent:     registration,
		AutoRenew: true,
	})
	if err != nil {
		return err
	}

	if response.LeaseId != "" {
		s.leases[agent.ID] = response.LeaseId
	}

	return nil
}

// List returns all registered agents
func (s *agentService) List(ctx context.Context) ([]*AgentInfo, error) {
	s.logger.Debug("Listing agents")

	client := generated.NewRegistryClient(s.client.conn)
	response, err := client.ListAgents(ctx, &generated.ListAgentsRequest{})
	if err != nil {
		return nil, err
	}

	agents := make([]*AgentInfo, 0, len(response.Agents))
	for _, agent := range response.Agents {
		agents = append(agents, agentFromRegistration(agent))
	}

	return agents, nil
}

// Get returns a specific agent by ID
func (s *agentService) Get(ctx context.Context, id string) (*AgentInfo, error) {
	s.logger.Debug("Getting agent", zap.String("id", id))

	client := generated.NewRegistryClient(s.client.conn)
	response, err := client.GetAgent(ctx, &generated.GetAgentRequest{AgentId: id})
	if err != nil {
		return nil, err
	}

	return agentFromRegistration(response), nil
}

// UpdateStatus updates an agent's status
func (s *agentService) UpdateStatus(ctx context.Context, id string, status AgentStatus) error {
	s.logger.Info("Updating agent status",
		zap.String("id", id),
		zap.String("status", string(status)),
	)

	return fmt.Errorf("update status is not supported by the registry API")
}

// UpdateConfidence updates an agent's confidence score
func (s *agentService) UpdateConfidence(ctx context.Context, id string, confidence float64) error {
	s.logger.Debug("Updating agent confidence",
		zap.String("id", id),
		zap.Float64("confidence", confidence),
	)
	
	if confidence < 0 || confidence > 1 {
		return fmt.Errorf("confidence must be between 0 and 1")
	}

	return fmt.Errorf("update confidence is not supported by the registry API")
}

// Heartbeat sends a heartbeat for an agent
func (s *agentService) Heartbeat(ctx context.Context, id string) error {
	s.logger.Debug("Sending heartbeat", zap.String("id", id))

	leaseId, ok := s.leases[id]
	if !ok {
		return fmt.Errorf("no lease found for agent %s", id)
	}

	client := generated.NewRegistryClient(s.client.conn)
	_, err := client.Renew(ctx, &generated.RenewRequest{LeaseId: leaseId})
	return err
}

// Unregister removes an agent from the system
func (s *agentService) Unregister(ctx context.Context, id string) error {
	s.logger.Info("Unregistering agent", zap.String("id", id))

	client := generated.NewRegistryClient(s.client.conn)
	_, err := client.Unregister(ctx, &generated.AgentRegistration{Id: id})
	if err == nil {
		delete(s.leases, id)
	}
	return err
}

// StreamAgents streams agent updates
func (s *agentService) StreamAgents(ctx context.Context) (<-chan *AgentInfo, error) {
	s.logger.Debug("Streaming agents")

	client := generated.NewRegistryClient(s.client.conn)
	stream, err := client.Watch(ctx, &generated.WatchRequest{
		IncludeInitial: true,
	})
	if err != nil {
		return nil, err
	}

	ch := make(chan *AgentInfo)
	go func() {
		defer close(ch)
		for {
			event, err := stream.Recv()
			if err != nil {
				return
			}
			if event.Agent == nil {
				continue
			}

			select {
			case ch <- agentFromRegistration(event.Agent):
			case <-ctx.Done():
				return
			}
		}
	}()

	return ch, nil
}

func agentFromRegistration(agent *generated.AgentRegistration) *AgentInfo {
	metadata := map[string]string{}
	if agent.GetMetadata() != nil {
		for key, value := range agent.Metadata.Labels {
			metadata[key] = value
		}
		if agent.Metadata.Version != "" {
			metadata["version"] = agent.Metadata.Version
		}
		if agent.Metadata.Region != "" {
			metadata["region"] = agent.Metadata.Region
		}
	}

	lastSeen := timeNow()
	if agent.RegisteredAt != nil {
		lastSeen = agent.RegisteredAt.AsTime()
	}

	defaultConfidence := 0.0
	if agent.GetMetadata() != nil {
		defaultConfidence = agent.Metadata.GetDefaultConfidence()
	}

	return &AgentInfo{
		ID:           agent.GetId(),
		Name:         agent.GetName(),
		Status:       AgentStatusActive,
		Capabilities: agent.GetCapabilities(),
		Endpoint:     agent.GetEndpoint(),
		LastSeen:     lastSeen,
		Confidence:   defaultConfidence,
		Metadata:     metadata,
	}
}

File: packages/sdk-rust/src/agent_service.rs
use crate::{
    error::{Error, Result},
    generated::parallax::registry::{
        registry_client::RegistryClient, AgentRegistration, GetAgentRequest, ListAgentsRequest,
        RegisterRequest, RenewRequest, WatchRequest,
    },
    types::{Agent, AgentStatus},
};
use futures::{Stream, StreamExt};
use std::pin::Pin;
use tonic::transport::Channel;
use tracing::{debug, info};

/// Service for agent operations
#[derive(Clone)]
pub struct AgentService {
    _channel: Channel,
}

impl AgentService {
    pub(crate) fn new(channel: Channel) -> Self {
        Self { _channel: channel }
    }

    /// Register a new agent
    pub async fn register(&self, mut agent: Agent) -> Result<Agent> {
        info!("Registering agent: {}", agent.name);
        
        // Ensure agent has an ID
        if agent.id.is_empty() {
            agent.id = uuid::Uuid::new_v4().to_string();
        }
        
        // Update last seen
        agent.last_seen = chrono::Utc::now();
        
        let metadata = agent.metadata.clone();
        let request = RegisterRequest {
            agent: Some(AgentRegistration {
                id: agent.id.clone(),
                name: agent.name.clone(),
                endpoint: agent.endpoint.clone(),
                capabilities: agent.capabilities.clone(),
                metadata: Some(crate::generated::parallax::registry::agent_registration::Metadata {
                    version: metadata.get("version").cloned().unwrap_or_default(),
                    region: metadata.get("region").cloned().unwrap_or_default(),
                    labels: metadata,
                    default_confidence: agent.confidence,
                }),
                registered_at: None,
                ttl: None,
            }),
            auto_renew: true,
        };

        let mut client = RegistryClient::new(self._channel.clone());
        client.register(request).await?;

        Ok(agent)
    }

    /// List all agents
    pub async fn list(&self) -> Result<Vec<Agent>> {
        debug!("Listing agents");

        let mut client = RegistryClient::new(self._channel.clone());
        let response = client
            .list_agents(ListAgentsRequest {
                capabilities: vec![],
                labels: Default::default(),
                limit: 0,
                continuation_token: String::new(),
            })
            .await?
            .into_inner();

        Ok(response
            .agents
            .into_iter()
            .map(agent_from_registration)
            .collect())
    }

    /// Get a specific agent
    pub async fn get(&self, id: &str) -> Result<Agent> {
        debug!("Getting agent: {}", id);
        
        let agents = self.list().await?;
        agents
            .into_iter()
            .find(|a| a.id == id)
            .ok_or_else(|| crate::error::Error::NotFound(format!("Agent not found: {}", id)))
    }

    /// Update agent status
    pub async fn update_status(&self, id: &str, status: AgentStatus) -> Result<()> {
        info!("Updating agent status: {} -> {:?}", id, status);

        Err(Error::InvalidArgument(format!(
            "update status is not supported by the registry API: {}",
            id
        )))
    }

    /// Update agent confidence
    pub async fn update_confidence(&self, id: &str, confidence: f64) -> Result<()> {
        debug!("Updating agent confidence: {} -> {}", id, confidence);
        
        if !(0.0..=1.0).contains(&confidence) {
            return Err(crate::error::Error::InvalidArgument(
                "Confidence must be between 0 and 1".to_string(),
            ));
        }
        

        Err(Error::InvalidArgument(format!(
            "update confidence is not supported by the registry API: {}",
            id
        )))
    }

    /// Send heartbeat for an agent
    pub async fn heartbeat(&self, id: &str) -> Result<()> {
        debug!("Sending heartbeat for agent: {}", id);

        let mut client = RegistryClient::new(self._channel.clone());
        client
            .renew(RenewRequest {
                lease_id: format!("lease-{}", id),
                ttl: None,
            })
            .await?;
        Ok(())
    }

    /// Unregister an agent
    pub async fn unregister(&self, id: &str) -> Result<()> {
        info!("Unregistering agent: {}", id);

        let mut client = RegistryClient::new(self._channel.clone());
        client
            .unregister(AgentRegistration {
                id: id.to_string(),
                name: String::new(),
                endpoint: String::new(),
                capabilities: Vec::new(),
                metadata: None,
                registered_at: None,
                ttl: None,
            })
            .await?;

        Ok(())
    }

    /// Stream agent updates
    pub async fn stream_agents(&self) -> Result<Pin<Box<dyn Stream<Item = Result<Agent>> + Send>>> {
        debug!("Streaming agents");

        let mut client = RegistryClient::new(self._channel.clone());
        let stream = client
            .watch(WatchRequest {
                capabilities: vec![],
                include_initial: true,
            })
            .await?
            .into_inner();

        let mapped = stream.filter_map(|event| async move {
            match event {
                Ok(event) => event.agent.map(|agent| Ok(agent_from_registration(agent))),
                Err(error) => Some(Err(Error::from(error))),
            }
        });

        Ok(Box::pin(mapped))
    }
}

fn agent_from_registration(agent: AgentRegistration) -> Agent {
    let metadata = agent
        .metadata
        .as_ref()
        .map(|metadata| metadata.labels.clone())
        .unwrap_or_default();
    let confidence = agent
        .metadata
        .as_ref()
        .map(|metadata| metadata.default_confidence)
        .unwrap_or(0.0);

    Agent {
        id: agent.id,
        name: agent.name,
        status: AgentStatus::Active,
        capabilities: agent.capabilities,
        endpoint: agent.endpoint,
        last_seen: chrono::Utc::now(),
        confidence,
        metadata,
    }
}
